source/Nevermore/Advanced/Queryable/QueryProvider.cs

```diff
@@ -86,7 +87,27 @@ public async Task<TResult> ExecuteAsync<TResult>(Expression expression, Cancella
             {
                 var asyncStream = (IAsyncEnumerable<object>)GenericStreamAsyncMethod.MakeGenericMethod(expression.Type)
                     .Invoke(queryExecutor, new object[] { command, cancellationToken });
-                return (TResult)Convert.ChangeType(await FirstOrDefaultAsync(asyncStream, cancellationToken), typeof(TResult));
+                var firstOrDefaultAsync = await FirstOrDefaultAsync(asyncStream, cancellationToken);
+
+                if (firstOrDefaultAsync is null)
+                {
+                    if (typeof(TResult).IsValueType)
+                    {
+                        throw new InvalidCastException("Null object cannot be converted to a value type");
```

Comment:
This is actually not what we want. `FirstOrDefault` should work with value types. Currently it can't, because the implementation of `FirstOrDefaultAsync` below only works with reference types. This is fine for now, because projection is not yet implemented, so you can only query complete documents, which must be classes.